src/clock_control/watchdog.rs

```diff
@@ -0,0 +1,227 @@
+//! RTC Watchdog implementation
+//!
+//! # TODO:
+//! - Add convenience methods for configuration
+//! - Consider add default configuration for start with time only
+
+use crate::prelude::*;
+use embedded_hal::watchdog::{Watchdog, WatchdogDisable, WatchdogEnable};
+use esp32::generic::Variant::Val;
+use esp32::rtccntl::wdtconfig0::*;
+use esp32::RTCCNTL;
+
+pub type WatchdogAction = WDT_STG0_A;
+pub type WatchDogResetDuration = WDT_CPU_RESET_LENGTH_A;
+
+const WATCHDOG_UNBLOCK_KEY: u32 = 0x50D83AA1;
+
+const WATCHDOG_BLOCK_VALUE: u32 = 0x89ABCDEF;
+
+pub struct WatchDog {
+    clock_control_config: super::ClockControlConfig,
+}
+
+/// Watchdog configuration
+#[derive(Debug)]
+pub struct WatchdogConfig<
+    T1: Into<MicroSeconds>,
+    T2: Into<MicroSeconds>,
+    T3: Into<MicroSeconds>,
+    T4: Into<MicroSeconds>,
+> {
+    pub period1: T1,
+    pub action1: WatchdogAction,
+    pub period2: T2,
+    pub action2: WatchdogAction,
+    pub period3: T3,
+    pub action3: WatchdogAction,
+    pub period4: T4,
+    pub action4: WatchdogAction,
+    pub cpu_reset_duration: WatchDogResetDuration,
+    pub sys_reset_duration: WatchDogResetDuration,
+    pub pause_in_sleep: bool,
+    pub reset_cpu: [bool; 2],
+}
+
+impl WatchDog {
+    /// internal function to create new watchdog structure
+    pub(crate) fn new(clock_control_config: super::ClockControlConfig) -> Self {
+        WatchDog {
+            clock_control_config,
+        }
+    }
+
+    /// function to unlock the watchdog (write unblock key) and lock after use
+    fn access_registers<A, F: FnMut(&esp32::rtccntl::RegisterBlock) -> A>(
+        &mut self,
+        mut f: F,
+    ) -> A {
+        // Unprotect write access to registers
+        let rtc_control = unsafe { &(*RTCCNTL::ptr()) };
+
+        rtc_control
+            .wdtwprotect
+            .write(|w| unsafe { w.bits(WATCHDOG_UNBLOCK_KEY) });
+
+        let a = f(rtc_control);
+
+        // Protect again
+        rtc_control
+            .wdtwprotect
+            .write(|w| unsafe { w.bits(WATCHDOG_BLOCK_VALUE) });
+
+        a
+    }
+
+    /// Calculate period from ref ticks
+    fn calc_period(&self, value: u32) -> MicroSeconds {
+        (((1000000u64 * value as u64)
+            / (u32::from(self.clock_control_config.slow_rtc_frequency()) as u64)) as u32)
+            .us()
+    }
+
+    /// Calculate ref ticks from period
+    fn calc_ticks(&self, value: MicroSeconds) -> u32 {
+        (u32::from(value) as u64 * u32::from(self.clock_control_config.slow_rtc_frequency()) as u64
+            / 1000000u64) as u32
+    }
+
+    /// Get watchdog configuration
+    pub fn config(
+        &self,
+    ) -> Result<WatchdogConfig<MicroSeconds, MicroSeconds, MicroSeconds, MicroSeconds>, super::Error>
```

Comment:
In the previous [comment](https://github.com/esp-rs/esp32-hal/pull/11/files/c10679e1e326bf8379fe8c8cc988873ea13c6491#r403485964) it makes this more readable, and anyway it will be using `MicroSeconds` under the hood.

Reply:
Correct. So depends on the outcome of our discussion on the previous item. (I care more about the public API, then the internal implementation.)

Reply:
Done